azurelinuxagent/ga/cgroupapi.py

```diff
@@ -342,11 +397,139 @@ def start_extension_command(self, extension_name, command, cmd_name, timeout, sh
             with self._systemd_run_commands_lock:
                 self._systemd_run_commands.remove(process.pid)
 
-    def cleanup_legacy_cgroups(self):
+
+class SystemdCgroupsApiv2(SystemdCgroupsApi):
+    """
+    Cgroups v2 interface via systemd
+    """
+
+    def is_controller_enabled(self, controller, cgroup_path):
         """
-        Previous versions of the daemon (2.2.31-2.2.40) wrote their PID to /sys/fs/cgroup/{cpu,memory}/WALinuxAgent/WALinuxAgent;
-        starting from version 2.2.41 we track the agent service in walinuxagent.service instead of WALinuxAgent/WALinuxAgent. If
-        we find that any of the legacy groups include the PID of the daemon then we need to disable data collection for this
-        instance (under systemd, moving PIDs across the cgroup file system can produce unpredictable results)
+        Returns True if the provided controller is enabled at the provided cgroup.
+
+        There are two ways to determine if a controller is enabled at the provided cgroup:
+
+        1. For non-leaf cgroups, the cgroup.subtree_control shows space separated list of the controllers which are
+        enabled to control resource distribution from the cgroup to its children. All non-root "cgroup.subtree_control"
+        files can only contain controllers which are enabled in the parent's "cgroup.subtree_control" file.
```

Comment:
As you stated for the given cgroup path, controllers enabled information present in parent cgroup.subtree_control file but you are checking in current path file and also, why do we need to check for every cgroup path. Isn't this check needed only for root?

Documentation stated this

`Top-down Constraint[¶](https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#top-down-constraint)
Resources are distributed top-down and a cgroup can further distribute a resource only if the resource has been distributed to it from the parent. This means that all non-root "cgroup.subtree_control" files can only contain controllers which are enabled in the parent's "cgroup.subtree_control" file. A controller can be enabled only if the parent has the controller enabled and a controller can't be disabled if one or more children have it enabled.`

Looks like root will have what controllers enabled, without here, you can't enable it in non-root cgroups

Reply:
If controller is in cgroup.subtree_control file then we know it is enabled in the current cgroup and any of its children due to top down constraint. That's why I had that check. If this is unclear then maybe cgroup.controllers is a better check for non-root cgroups, and this method would be better divided by root vs non-root instead of leaf vs non-leaf.

Answer to the second question in your other comments

Reply:
For example, if subtree_contorl is empty in agent service cgroup directory but controllers enabled in azure.slice, then this fails and says that controllers are not enabled when it's enabled

Reply:
It won't fail since it checks for existence of <controller>.* interface files, which only exist if controller is enabled

Reply:
I agree cgroup.controllers would be clearer check though

Reply:
I just want to make sure what I speak correct here. WhenI check doc for cgroup.controller, it just says what controllers available to enable but it does not say what enabled. 

`Each cgroup has a "cgroup.controllers" file which lists all controllers available for the cgroup to enable:
`
If we want to use this, then we need to check is this file containing all the controllers minus what's mounted in v1? or will it show everything regardless v1 status?

Reply:
If there is a controller (cpu for example) mounted in v1, then it will not be mounted in v2. Controllers cannot be mounted in both hierarchies simultaneously, so it wouldn't appear in the list of available controllers for v2. 

cgroup.controllers file lists all controllers available for the cgroup to enable to its children. A cgroup can only enable a controller to its children if it is enabled in the current cgroup itself, due to top down constraint.

Reply:
I am not sure we need this check other than for the root cgroup. I think we need to check whether cpu & memory are enabled (cgroup.subtree_control) on root. If they are enabled, we should use them. If not. we should not enable them. at least in the initial releases -- we can reconsider this later.

We own the non-root cgroups (walinuxagent.service, azure-walinuxagent-logcollector.slice, azure-vmextensions.slice, azure-vmextensions-Extension.Foo-1.0.0.0.slice) and we should enable the controllers we want to use on those cgroups.